File: WdtcCore/src/main/java/org/wdt/wdtc/core/launch/GameLibraryData.java
package org.wdt.wdtc.core.launch;


import lombok.SneakyThrows;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.wdtc.core.download.infterface.DownloadSourceInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.LibraryObject;
import org.wdt.wdtc.core.manger.DownloadSourceManger;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;

public class GameLibraryData {
    private final Launcher launcher;
    private final DownloadSourceInterface source;


    public GameLibraryData(Launcher launcher) {
        this.launcher = launcher;
        this.source = DownloadSourceManger.getDownloadSource();
    }

    public File GetNativesLibraryFile(LibraryObject.NativesOs nativesOs) {
        return new File(launcher.getGameLibraryDirectory(), nativesOs.getPath());
    }

    @SneakyThrows(MalformedURLException.class)
    public URL GetNativesLibraryUrl(LibraryObject libraryObject) {
        LibraryObject.NativesOs Nativesindows = libraryObject.getDownloads().getClassifiers().getNativesindows();
        if (DownloadSourceManger.isNotOfficialDownloadSource()) {
            return new URL(source.getLibraryUrl() + Nativesindows.getPath());
        } else {
            return Nativesindows.getUrl();
        }
    }

    public File GetLibraryFile(LibraryObject object) {
        File LibraryPath = launcher.getGameLibraryDirectory();
        DependencyDownload dependency = new DependencyDownload(object.getLibraryName());
        dependency.setDownloadPath(LibraryPath);
        return dependency.getLibraryFile();

    }

    @SneakyThrows(MalformedURLException.class)
    public URL GetLibraryUrl(LibraryObject libraryObject) {
        if (DownloadSourceManger.isNotOfficialDownloadSource()) {
            DependencyDownload dependency = new DependencyDownload(libraryObject.getLibraryName());
            dependency.setDefaultUrl(source.getLibraryUrl());
            return dependency.getLibraryUrl();
        } else {
            return GetOfficialLibraryUrl(libraryObject);
        }
    }

    public URL GetOfficialLibraryUrl(LibraryObject object) {
        return object.getDownloads().getArtifact().getUrl();
    }

}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/launch/GameLibraryData.kt
package org.wdt.wdtc.core.launch

import org.wdt.wdtc.core.game.Launcher
import org.wdt.wdtc.core.game.LibraryObject
import org.wdt.wdtc.core.game.LibraryObject.NativesOs
import org.wdt.wdtc.core.manger.downloadSource
import org.wdt.wdtc.core.manger.isNotOfficialDownloadSource
import org.wdt.wdtc.core.utils.dependency.DependencyDownload
import java.io.File
import java.net.URL

open class GameLibraryData(private val launcher: Launcher) {

  fun getNativesLibraryFile(nativesOs: NativesOs): File {
    return File(launcher.gameLibraryDirectory, nativesOs.path!!)
  }

  fun getNativesLibraryUrl(libraryObject: LibraryObject): URL {
    val nativesindows = libraryObject.downloads?.classifiers?.nativesindows
    return if (isNotOfficialDownloadSource) URL(downloadSource.libraryUrl + nativesindows?.path) else nativesindows?.url!!
  }

  fun getLibraryFile(libraryObject: LibraryObject): File {
    val libraryPath = launcher.gameLibraryDirectory
    val dependency = DependencyDownload(libraryObject.libraryName!!)
    dependency.downloadPath = libraryPath
    return dependency.libraryFile
  }

  fun getLibraryUrl(libraryObject: LibraryObject): URL {
    return if (isNotOfficialDownloadSource) {
      val dependency = DependencyDownload(libraryObject.libraryName!!)
      dependency.defaultUrl = downloadSource.libraryUrl
      dependency.libraryUrl
    } else {
      getOfficialLibraryUrl(libraryObject)
    }
  }

  fun getOfficialLibraryUrl(libraryObject: LibraryObject): URL {
    return libraryObject.downloads?.artifact?.url!!
  }
}
